xapian-applications/omega/index_file.cc

```diff
@@ -545,6 +545,77 @@ index_add_document(const string & urlterm, time_t last_altered,
     }
 }
 
+bool
+index_update_entry(const string & urlterm,
+		   DirectoryIterator & d,
+		   Xapian::docid & did)
+{
+    if (!did) {
+	Xapian::PostingIterator p = db.postlist_begin(urlterm);
+	if (p != db.postlist_end(urlterm))
+	    did = *p;
+    }
+
+    if (did) {
+	time_t mtime = d.get_mtime();
+	time_t ctime = d.get_ctime();
+	Xapian::Document doc = db.get_document(did);
+	string value_ctime = doc.get_value(VALUE_CTIME);
+	time_t doc_ctime = binary_string_to_int(value_ctime);
+	string value_mtime = doc.get_value(VALUE_LASTMOD);
+	time_t doc_mtime = binary_string_to_int(value_mtime);
+
+	if (doc_mtime == mtime && doc_ctime <= ctime) {
+	    if (verbose)
+		cout << "Updating entry without re-extracting text" << endl;
+	    // Remove terms from the document
+	    Xapian::TermIterator term_iterator = doc.termlist_begin();
+	    while (term_iterator != doc.termlist_end()) {
+		const string& term = *term_iterator;
+		bool rem_term = startswith(term, "I") ||
+				startswith(term, "O");
+		// I: boolean filter term for "can see" permission
+		// O: Owner
+		if (rem_term) {
+		    doc.remove_term(term);
+		}
+		term_iterator++;
+	    }
+	    // Updating document terms
+	    bool inc_tag_added = false;
+	    if (d.is_other_readable()) {
+		inc_tag_added = true;
+		doc.add_boolean_term("I*");
+	    } else if (d.is_group_readable()) {
+		const char * group = d.get_group();
+		if (group) {
+		    doc.add_boolean_term(string("I#") + group);
+		}
+	    }
+
+	    const char * owner = d.get_owner();
+	    if (owner) {
+		doc.add_boolean_term(string("O") + owner);
+		if (!inc_tag_added && d.is_owner_readable())
+		    doc.add_boolean_term(string("I@") + owner);
+	    }
```

Comment:
This looks like a cut-and-paste of ~18 lines of existing code from `index_mimetype()` - probably better to split this out into a new function and call it from here and `index_mimetype()`.

Reply:
I was not sure about creating a new function. I'm going to add it.